ocaml/xapi/repository_helpers.ml

```diff
@@ -99,54 +99,35 @@ module GuidanceSet = struct
   include GuidanceSet'
   open Guidance
 
-  let eq l s = equal (of_list l) (of_list s)
-
-  let eq_set1 = eq [EvacuateHost; RestartToolstack]
-
-  let eq_set2 = eq [RestartDeviceModel; RestartToolstack]
-
-  let error_msg l =
-    Printf.sprintf "Found wrong guidance(s): %s"
-      (String.concat ";" (List.map to_string l))
-
-  let assert_valid_guidances = function
-    | []
-    | [RebootHost]
-    | [EvacuateHost]
-    | [RestartToolstack]
-    | [RestartDeviceModel] ->
-        ()
-    | l when eq_set1 l ->
-        (* EvacuateHost and RestartToolstack *)
-        ()
-    | l when eq_set2 l ->
-        (* RestartDeviceModel and RestartToolstack *)
-        ()
-    | l ->
-        let msg = error_msg l in
-        raise Api_errors.(Server_error (internal_error, [msg]))
-
   let precedences =
     [
       (RebootHost, of_list [RestartToolstack; EvacuateHost; RestartDeviceModel])
     ; (EvacuateHost, of_list [RestartDeviceModel])
+    ; (RestartVM, of_list [RestartDeviceModel])
     ]
 
-  let resort_guidances ~remove_evacuations gs =
-    let gs' =
-      List.fold_left
-        (fun acc (higher, lowers) ->
-          if mem higher acc then
-            diff acc lowers
-          else
-            acc
-        )
-        gs precedences
-    in
-    if remove_evacuations then
-      remove EvacuateHost gs'
-    else
-      gs'
+  let resort gs =
+    List.fold_left
+      (fun acc (higher, lowers) ->
+        if mem higher acc then
+          diff acc lowers
+        else
+          acc
+      )
+      gs precedences
+
+  (** remove the guidances which are in s1 from s2. *)
+  let reduce s1 s2 = diff (resort (union s1 s2)) s1
+
+  (** for each set in the list, remove the guidances which are in previous sets from it. *)
```

Comment:
Are we dealing with lists (order is important) or sets (order is not important)? If we are operating on sets I would prefer sets being used.

Reply:
In this module, only `set`s are used. The `list` here is for a list of sets.